common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -20,150 +20,420 @@
         string = null;
-        caseInsensitiveHash = 0;
-        super.arrayChanged();
+        hash = 0;
     }
 
-    @Override
-    public String toString(Charset charset, int start, int end) {
-        if (start == 0 && end == length()) {
-            if (string == null) {
-                string = super.toString(charset, start, end);
-            }
-            return string;
+    /**
+     * This gives direct access to the underlying storage array.
+     * The {@link #toByteArray()} should be preferred over this method.
+     * If the return value is changed then {@link #arrayChanged()} must be called.
+     * @see #arrayOffset()
+     * @see #isEntireArrayUsed()
+     */
+    public byte[] array() {
+        return value;
+    }
+
+    /**
+     * The offset into {@link #array()} for which data for this ByteString begins.
+     * @see #array()
+     * @see #isEntireArrayUsed()
+     */
+    public int arrayOffset() {
+        return offset;
+    }
+
+    /**
+     * Determine if the storage represented by {@link #array()} is entirely used.
+     * @see #array()
+     */
+    public boolean isEntireArrayUsed() {
+        return offset == 0 && length == value.length;
+    }
+
+    /**
+     * Converts this string to a byte array.
+     */
+    public byte[] toByteArray() {
+        return toByteArray(0, length());
+    }
+
+    /**
+     * Converts a subset of this string to a byte array.
+     * The subset is defined by the range [{@code start}, {@code end}).
+     */
+    public byte[] toByteArray(int start, int end) {
+        return Arrays.copyOfRange(value, start + offset, end + offset);
+    }
+
+    /**
+     * Copies the content of this string to a byte array.
+     *
+     * @param srcIdx the starting offset of characters to copy.
+     * @param dst the destination byte array.
+     * @param dstIdx the starting offset in the destination byte array.
+     * @param length the number of characters to copy.
+     */
+    public void copy(int srcIdx, byte[] dst, int dstIdx, int length) {
+        if (isOutOfBounds(srcIdx, length, length())) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= srcIdx(" + srcIdx + ") <= srcIdx + length("
+                            + length + ") <= srcLen(" + length() + ')');
         }
 
-        return super.toString(charset, start, end);
+        System.arraycopy(value, srcIdx + offset, checkNotNull(dst, "dst"), dstIdx, length);
+    }
+
+    @Override
+    public char charAt(int index) {
+        return b2c(byteAt(index));
+    }
+
+    /**
+     * Determines if this {@code String} contains the sequence of characters in the {@code CharSequence} passed.
+     *
+     * @param cs the character sequence to search for.
+     * @return {@code true} if the sequence of characters are contained in this string, otherwise {@code false}.
+     */
+    public boolean contains(CharSequence cs) {
+        if (cs == null) {
+            throw new NullPointerException();
```

Comment:
also if indexOf(...) already check for null you may just remove this check here.